packages/core/src/WorkflowExecute.ts

```diff
@@ -1195,6 +1209,113 @@ export class WorkflowExecute {
 						taskData,
 						this.runExecutionData,
 					]);
+
+					let nodeNames: string[] = Object.keys(
+						this.runExecutionData.executionData!.waitingExecution,
+					);
+
+					if (
+						workflow.version === 2 &&
+						this.runExecutionData.executionData!.nodeExecutionStack.length === 0 &&
+						nodeNames.length
+					) {
+						// There are no more nodes in the execution stack. So would normally stop now.
+						// For workflows of version 2, go however through and execute the waiting nodes
+						// with one missing input, one by one.
+
+						for (let i = 0; i < nodeNames.length; i++) {
+							const nodeName = nodeNames[i];
+
+							const node = workflow.getNode(nodeName);
+							if (!node) {
+								continue;
+							}
+							const nodeType = workflow.nodeTypes.getByNameAndVersion(
+								node.type,
+								node.typeVersion,
+							) as INodeType;
+
+							// Check if the node is only allowed execute if both inputs received data
+							let allInputsRequired = nodeType.description.allInputsRequired;
+							if (allInputsRequired !== undefined) {
+								if (typeof allInputsRequired === 'string') {
+									allInputsRequired = !!workflow.expression.getSimpleParameterValue(
+										node,
+										allInputsRequired,
+										this.mode,
+										this.additionalData.timezone,
+										{},
+									);
+								}
+
+								if (allInputsRequired) {
+									continue;
+								}
+							}
+
+							const parentNodes = workflow.getParentNodes(nodeName);
+
+							// Check if input nodes (of same run) got already executed
+							// eslint-disable-next-line @typescript-eslint/no-loop-func
+							const found = parentNodes.some((value) => nodeNames.includes(value));
+							if (found) {
+								// Execute node later as one of its dependencies is still outstanding
+								continue;
+							}
+
+							const runIndexes = Object.keys(
+								this.runExecutionData.executionData!.waitingExecution[nodeName],
+							).sort();
+
+							const firstRunIndex = parseInt(runIndexes[0]);
```

Comment:
Are we certain that runIndexes.length is always > 0 here?

Reply:
Yes to 100%. After all, do is `runIndexes` generated from `this.runExecutionData.executionData.waitingExecution` and data from it is what we iterate over.